### PR TITLE
Feat/optimalization

### PR DESCRIPTION
**Pull request by unkn0wn-root**

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -87,8 +89,11 @@ type InMemoryCache[K comparable, V any] struct {
 	cleanupCh   chan struct{}
 	closeCh     chan struct{}
 	closeOnce   sync.Once
-	closed      int32     // Atomic flag indicating cache is closed
-	itemPool    sync.Pool // Object pool for cache items
+	closed      int32     // Flag indicating cache is closed
+	itemPool    sync.Pool // Object pool
+
+	cachedTime   int64         // Cached timestamp updated by background goroutine
```
Comment (ccoVeille) [inline cache.go:95]: Did you consider using [atomic.Int64](https://pkg.go.dev/sync/atomic#Int64)

These types were added with Go 1.19 to simplify the usage of atomic

File: cache.go [hunk 2]
```diff
@@ -13,6 +13,8 @@ import (
 const (
 	NoExpiration      time.Duration = -1 // Never expires
 	DefaultExpiration time.Duration = 0  // Use default config value
+
+	CachedTimeStampUpdaterIntervals = 100 * time.Millisecond // How often should updater run
```
Comment (ccoVeille) [inline cache.go:17]: Do you need this variable, and other to be exported ?

I mean, do you consider people would use `kioshun.CachedTimeStampUpdaterIntervals = delay` ?

File: cache.go [hunk 3]
```diff
@@ -433,30 +482,43 @@ func (c *InMemoryCache[K, V]) Exists(key K) bool {
...
+	if item.expireTime == 0 {
+		shard.mu.RUnlock()
+		return true
+	}
+
+	// For expiring items, check expiration with cached timestamp
+	now := atomic.LoadInt64(&c.cachedTime)
+	if now > item.expireTime {
+		shard.mu.RUnlock()
+		shard.mu.Lock()
```
Comment (ccoVeille) [inline cache.go:503]: I would have inverted them to reduce the risk of having the resource getting unlocked

But I'm unsure it works

File: cache.go [hunk 4]
```diff
@@ -433,30 +482,43 @@ func (c *InMemoryCache[K, V]) Exists(key K) bool {
...
-		if c.config.EvictionPolicy == LFU && item.heapIndex != -1 {
-			heap.Remove(shard.lfuHeap, item.heapIndex)
-		}
-		c.itemPool.Put(item)
-		atomic.AddInt64(&shard.size, -1)
-		if c.config.StatsEnabled {
-			atomic.AddInt64(&shard.expirations, 1)
+	// Fast path for non-expiring items
+	if item.expireTime == 0 {
+		shard.mu.RUnlock()
```
Comment (ccoVeille) [inline cache.go:495]: I feel like you are unlocking in all branch.

Maybe a defer before this if ?

**Comment by unkn0wn-root** [comment]
After making more testing, there are some edge cases and some potential race conditions with SetWithCallback so I'm closing this PR for now.
